internal/page/donor/your_name.go

```diff
@@ -0,0 +1,131 @@
+package donor
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+
+	"github.com/gorilla/sessions"
+	"github.com/ministryofjustice/opg-go-common/template"
+	"github.com/ministryofjustice/opg-modernising-lpa/internal/actor"
+	"github.com/ministryofjustice/opg-modernising-lpa/internal/page"
+	"github.com/ministryofjustice/opg-modernising-lpa/internal/sesh"
+	"github.com/ministryofjustice/opg-modernising-lpa/internal/validation"
+)
+
+type yourNameData struct {
+	App         page.AppData
+	Errors      validation.List
+	Form        *yourNameForm
+	NameWarning *actor.SameNameWarning
+}
+
+func YourName(tmpl template.Template, donorStore DonorStore, sessionStore sessions.Store) Handler {
+	return func(appData page.AppData, w http.ResponseWriter, r *http.Request, donor *actor.DonorProvidedDetails) error {
+		data := &yourNameData{
+			App: appData,
+			Form: &yourNameForm{
+				FirstNames: donor.Donor.FirstNames,
+				LastName:   donor.Donor.LastName,
+				OtherNames: donor.Donor.OtherNames,
+			},
+		}
+
+		if r.Method == http.MethodGet && data.Form.FirstNames == "" {
+			if latestDonor, _ := donorStore.Latest(r.Context()); latestDonor != nil {
+				data.Form = &yourNameForm{
+					FirstNames: latestDonor.Donor.FirstNames,
+					LastName:   latestDonor.Donor.LastName,
+					OtherNames: latestDonor.Donor.OtherNames,
+				}
+			}
+		}
+
+		if r.Method == http.MethodPost {
+			loginSession, err := sesh.Login(sessionStore, r)
+			if err != nil {
+				return err
+			}
+			if loginSession.Email == "" {
+				return fmt.Errorf("no email in login session")
+			}
+
+			data.Form = readYourNameForm(r)
+			data.Errors = data.Form.Validate()
+
+			nameWarning := actor.NewSameNameWarning(
+				actor.TypeDonor,
+				donorMatches(donor, data.Form.FirstNames, data.Form.LastName),
+				data.Form.FirstNames,
+				data.Form.LastName,
+			)
+
+			if data.Errors.Any() ||
+				data.Form.IgnoreNameWarning != nameWarning.String() &&
+					donor.Donor.FullName() != fmt.Sprintf("%s %s", data.Form.FirstNames, data.Form.LastName) {
+				data.NameWarning = nameWarning
+			}
+
+			if !data.Errors.Any() && data.NameWarning == nil {
+				changesMade := donor.Donor.FirstNames != data.Form.FirstNames || donor.Donor.LastName != data.Form.LastName
```

Comment:
just to check, are you purposefully not checking other names here?

Reply:
also, this doesn't work if you do:
- create registered LPA in fixtures
- go to create a new LPA
- go to change name
- don't edit the name, just continue
- see the "we have updated your details" page (but nothing has changed...)

Reply:
and should there be a similar check on the date of birth and address pages?

Reply:
Totally had the intentions of adding to the other handlers - should be working now